src/backend/joanie/core/api/client.py

```diff
@@ -71,79 +74,133 @@ class CourseProductRelationViewSet(
-            except KeyError as exc:
-                raise DRFValidationError(
-                    {
-                        "course": _(
-                            "You must specify a course code to get product details."
-                        )
-                    }
-                ) from exc
+        return super().retrieve(request, *args, **kwargs)
 
-            queryset = queryset.filter(
-                course_relations__course__code=course_code,
-                course_relations__organizations__isnull=False,
-            )
-        else:
-            queryset = queryset.filter(course_relations__isnull=False)
+    def retrieve(self, request, *args, **kwargs):
+        """
+        The retrieve action is used to get a single course product relation.
+        The response is cached per language.
+        """
+        if self.kwargs.get("course_id"):
+            return self.retrieve_through_nested_course(request, *args, **kwargs)
 
-        return queryset.select_related("certificate_definition").distinct()
+        return super().retrieve(request, *args, **kwargs)
 
-    def get_serializer_context(self):
+    @property
+    def username(self):
+        """Get the authenticated username from the request."""
+        return (
+            self.request.auth["username"]
+            if self.request.auth
+            else self.request.user.username
+        )
+
+    def get_queryset(self):
         """
-        Provide course code to the `ProductSerializer`
+        The queryset filter depends on the action as to list course product relation we
+        only want to list course product relation for which the user has access to.
         """
-        context = super().get_serializer_context()
+        queryset = super().get_queryset()
 
-        if course := self.request.query_params.get("course"):
-            context.update({"course_code": course})
+        if course_id := self.kwargs.get("course_id"):
+            queryset = queryset.filter(**{self.course_lookup_filter: course_id})
 
-        return context
+        if organization_id := self.kwargs.get("organization_id"):
+            user_role_query = models.OrganizationAccess.objects.filter(
+                organization__accesses__user__username=self.username
+            ).values("role")[:1]
+            queryset = queryset.filter(
+                organizations__id=organization_id,
+                organizations__accesses__user__username=self.username,
+            ).annotate(user_role=Subquery(user_role_query))
+        elif self.action == "list" or self.action == "retrieve" and not course_id:
+            user_role_query = models.CourseAccess.objects.filter(
+                course__accesses__user__username=self.username
+            ).values("role")[:1]
+            queryset = queryset.filter(
+                course__accesses__user__username=self.username
+            ).annotate(user_role=Subquery(user_role_query))
```

Comment:
Hmmm... this is required for internal logic from Organization and Course models. Furthermore, from my point of view, it appears more logic to talk of user_role for the corresponding resource instead of resource role for the authenticated user.